src/Storages/MergeTree/MergeTreeDataSelectExecutor.cpp

```diff
@@ -1237,144 +1237,200 @@ Pipe MergeTreeDataSelectExecutor::spreadMarkRangesAmongStreamsFinal(
-    {
-        switch (data.merging_params.mode)
+        /// If do_not_merge_across_partitions_select_final is true and there is only one part in partition
+        /// then we won't postprocess this part
+        if (data_settings->do_not_merge_across_partitions_select_final &&
+            std::distance(parts_to_merge_ranges[range_index], parts_to_merge_ranges[range_index + 1]) == 1)
         {
-            case MergeTreeData::MergingParams::Ordinary:
-            {
-                return std::make_shared<MergingSortedTransform>(header, pipe.numOutputPorts(),
-                           sort_description, max_block_size);
-            }
+            partition_pipes.emplace_back(std::move(pipe));
+            continue;
+        }
 
-            case MergeTreeData::MergingParams::Collapsing:
-                return std::make_shared<CollapsingSortedTransform>(header, pipe.numOutputPorts(),
-                           sort_description, data.merging_params.sign_column, true, max_block_size);
+        pipe.addSimpleTransform([&metadata_snapshot](const Block & header)
+        {
+            return std::make_shared<ExpressionTransform>(header, metadata_snapshot->getSortingKey().expression);
+        });
 
-            case MergeTreeData::MergingParams::Summing:
-                return std::make_shared<SummingSortedTransform>(header, pipe.numOutputPorts(),
-                           sort_description, data.merging_params.columns_to_sum, partition_key_columns, max_block_size);
+        Names sort_columns = metadata_snapshot->getSortingKeyColumns();
+        SortDescription sort_description;
+        size_t sort_columns_size = sort_columns.size();
+        sort_description.reserve(sort_columns_size);
 
-            case MergeTreeData::MergingParams::Aggregating:
-                return std::make_shared<AggregatingSortedTransform>(header, pipe.numOutputPorts(),
-                           sort_description, max_block_size);
+        Names partition_key_columns = metadata_snapshot->getPartitionKey().column_names;
 
-            case MergeTreeData::MergingParams::Replacing:
-                return std::make_shared<ReplacingSortedTransform>(header, pipe.numOutputPorts(),
-                           sort_description, data.merging_params.version_column, max_block_size);
+        Block header = pipe.getHeader();
+        for (size_t i = 0; i < sort_columns_size; ++i)
+            sort_description.emplace_back(header.getPositionByName(sort_columns[i]), 1, 1);
 
-            case MergeTreeData::MergingParams::VersionedCollapsing:
-                return std::make_shared<VersionedCollapsingTransform>(header, pipe.numOutputPorts(),
-                           sort_description, data.merging_params.sign_column, max_block_size);
+        auto get_merging_processor = [&]() -> MergingTransformPtr
+        {
+            switch (data.merging_params.mode)
+            {
+                case MergeTreeData::MergingParams::Ordinary: {
```

Comment:
Style. According to out style guide, curly bracket should be on new line or can be ommitted.